py-polars/polars/utils/udfs.py

```diff
@@ -136,6 +131,53 @@ def _get_param_name(function: Callable[[Any], Any]) -> str | None:
         except ValueError:
             return None
 
+    def _inject_nesting(
+        self,
+        expression_blocks: dict[int, str],
+        logical_instructions: list[Instruction],
+    ) -> list[tuple[int, str]]:
+        """Inject nesting boundaries into expression blocks (as parentheses)."""
+        if logical_instructions:
+            # reconstruct nesting boundaries for mixed and/or ops by associating
+            # control flow jump offsets with their target expression blocks and
+            # injecting appropriate parentheses
+            combined_offset_idxs = set()
+            if len({inst.opname for inst in logical_instructions}) > 1:
+                block_offsets: list[int] = list(expression_blocks.keys())
+                previous_logical_opname = ""
+                for i, inst in enumerate(logical_instructions):
+                    # operator precedence means that we can combine logically connected
+                    # 'and' blocks into one (depending on follow-on logic) and should
+                    # parenthesise nested 'or' blocks
+                    logical_op = OpNames.CONTROL_FLOW[inst.opname]
+                    start = block_offsets[bisect_left(block_offsets, inst.offset) - 1]
+                    if previous_logical_opname == (
+                        "POP_JUMP_FORWARD_IF_FALSE"
+                        if _MIN_PY311
+                        else "POP_JUMP_IF_FALSE"
+                    ):
+                        # combine logical '&' blocks (and update start/block_offsets)
+                        prev = block_offsets[bisect_left(block_offsets, start) - 1]
+                        expression_blocks[prev] += f" & {expression_blocks.pop(start)}"
+                        block_offsets = list(expression_blocks.keys())
```

Comment:
this could just be `block_offsets.remove(start)`?

Reply:
(Yup - and that sound you just heard? That was my palm hitting my forehead 😜)